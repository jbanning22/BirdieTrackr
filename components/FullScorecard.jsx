import {
  StyleSheet,
  Text,
  View,
  Button,
  FlatList,
  TouchableOpacity,
} from 'react-native';
import React, {useState, useEffect} from 'react';
import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {SafeAreaView} from 'react-native-safe-area-context';
// import {Dimensions} from 'react-native';

const FullScorecard = ({route, navigation}) => {
  const {id} = route.params;
  const [scorecardData, setScorecardData] = useState({});
  const [holesData, setHolesData] = useState([]);
  const [showEndButton, setShowEndButton] = useState(false);

  const getScorecard = async () => {
    const token = await AsyncStorage.getItem('token');
    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const scoreC = await axios.get(
        `http://192.168.1.154:3000/scorecard/${id}`,
        {
          headers,
        },
      );
      act(() => {
        setScorecardData(scoreC.data);
      });
      act(() => {
        setHolesData(scoreC.data.holes);
      });
    } catch (error) {
      console.log(error);
    }
  };

  const updateStrokesPlus = async (id, strokes) => {
    const token = await AsyncStorage.getItem('token');

    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const updatedHoleP = await axios.patch(
        `http://192.168.1.154:3000/hole/${id}`,
        {strokes: strokes + 1},
        {headers},
      );
      getScorecard();
    } catch (error) {
      console.log(error);
    }
  };

  const updateStrokesMinus = async (id, strokes) => {
    const token = await AsyncStorage.getItem('token');

    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const updatedHoleM = await axios.patch(
        `http://192.168.1.154:3000/hole/${id}`,
        {strokes: strokes - 1},
        {headers},
      );
      getScorecard();
    } catch (error) {
      console.log(error);
    }
  };

  const updateParPlus = async (id, par) => {
    const token = await AsyncStorage.getItem('token');

    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const updatedHolePar = await axios.patch(
        `http://192.168.1.154:3000/hole/${id}`,
        {par: par + 1},
        {headers},
      );
      getScorecard();
    } catch (error) {
      console.log(error);
    }
  };
  const updateParMinus = async (id, par) => {
    const token = await AsyncStorage.getItem('token');

    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const updatedHoleParM = await axios.patch(
        `http://192.168.1.154:3000/hole/${id}`,
        {par: par - 1},
        {headers},
      );
      getScorecard();
    } catch (error) {
      console.log(error);
    }
  };

  const finishScorecard = async () => {
    const scorecardId = scorecardData.id;
    const token = await AsyncStorage.getItem('token');
    const headers = {
      Authorization: `Bearer ${token}`,
    };
    try {
      const finishCard = await axios.patch(
        `http://192.168.1.154:3000/scorecard/${scorecardId}`,
        {isCompleted: true, courseLength: scorecardData.courseLength},
        {headers},
      );
      navigation.navigate('Scorecard');
    } catch (error) {
      console.log(error);
    }
  };

  const renderItem = ({item}) => {
    return (
      <View style={styles.flatlistView}>
        <Text style={styles.renderParentText}>Hole {item.holeNumber}</Text>
        <View style={styles.parStrokeView}>
          <Text style={styles.renderText}>Par: {item.par}</Text>
          <View style={{flexDirection: 'column'}}>
            <TouchableOpacity
              style={styles.parButton}
              activeOpacity={0.5}
              onPress={() => updateParPlus(item.id, item.par)}>
              <Text style={styles.buttonTextPlus}>+</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.parButton}
              activeOpacity={0.5}
              onPress={() => updateParMinus(item.id, item.par)}>
              <Text style={styles.buttonTextMinus}>-</Text>
            </TouchableOpacity>
          </View>
        </View>
        <View style={styles.parStrokeView}>
          <Text style={styles.renderText}>Strokes: {item.strokes}</Text>
          <View style={{flexDirection: 'column'}}>
            <TouchableOpacity
              style={styles.parButton}
              activeOpacity={0.5}
              onPress={() => updateStrokesPlus(item.id, item.strokes)}>
              <Text style={styles.buttonTextPlus}>+</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.parButton}
              activeOpacity={0.5}
              onPress={() => updateStrokesMinus(item.id, item.strokes)}>
              <Text style={styles.buttonTextMinus}>-</Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
    );
  };

  useEffect(() => {
    getScorecard();
  }, []);

  const handleEndReached = () => {
    setShowEndButton(true);
  };

  const sortedData = holesData.sort((a, b) => a.holeNumber - b.holeNumber);
  return (
    <SafeAreaView style={styles.box1}>
      <Text style={styles.homeText}>{scorecardData.courseName} Scorecard</Text>
      <View style={styles.flatlistContainer}>
        <FlatList
          renderItem={renderItem}
          data={sortedData}
          showsHorizontalScrollIndicator={false}
          horizontal={true}
          pagingEnabled={true}
          onEndReached={handleEndReached}
          snapToAlignment={'center'}
        />
      </View>
      {showEndButton && (
        <TouchableOpacity
          style={styles.scorecardButton}
          onPress={finishScorecard}>
          <Text style={styles.scorecardButtonText}>Finish Scorecard</Text>
        </TouchableOpacity>
      )}
      <TouchableOpacity
        activeOpacity={0.5}
        onPress={() => navigation.navigate('Scorecard')}>
        <Text style={{alignSelf: 'center', marginTop: 50}}>Back</Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default FullScorecard;

const styles = StyleSheet.create({
  box1: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-start',
    backgroundColor: '#DDDDDD',
  },
  homeText: {
    fontSize: 44,
    fontWeight: '400',
    fontFamily: 'Helvetica',
    textAlign: 'center',
    color: '#DB6F52',
  },
  parButton: {
    height: 25,
    width: 25,
  },
  buttonTextPlus: {
    alignSelf: 'center',
    justifyContent: 'center',
    fontSize: 24,
    fontWeight: '600',
    color: '#13E12B',
  },
  buttonTextMinus: {
    alignSelf: 'center',
    justifyContent: 'center',
    fontSize: 24,
    fontWeight: '600',
    color: '#EA5C09',
  },
  scorecardButton: {
    width: 150,
    height: 40,
    borderRadius: 25,
    backgroundColor: '#52BEDB',
    justifyContent: 'center',
    padding: 5,
    // marginLeft: 150,
    // marginTop: 20,
  },
  scorecardButtonText: {
    color: 'white',
    fontWeight: '500',
    alignSelf: 'center',
    fontSize: 16,
  },
  parStrokeView: {
    flexDirection: 'row',
    alignContent: 'center',
    justifyContent: 'center',
  },
  flatlistContainer: {
    flex: 0,
    width: '100%',
    // marginTop: 40,
  },
  renderParentText: {
    alignSelf: 'center',
    fontSize: 20,
    fontWeight: '600',
    fontSize: 22,
    margin: 25,
  },
  renderText: {
    alignSelf: 'center',
    fontSize: 16,
    fontWeight: '300',
    fontSize: 22,
    margin: 15,
  },
  flatlistView: {
    height: 400,
    width: 395,
    alignContent: 'center',
    justifyContent: 'center',
    // borderRightWidth: 2,
    // borderColor: 'black',
    padding: 10,
    marginTop: 20,
  },
});
