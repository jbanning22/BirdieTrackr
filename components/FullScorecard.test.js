import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react-native';
import FullScorecard from './FullScorecard';

describe('Full Scorecard Screen', () => {
  it('should match the snapshot', () => {
    const route = {
      params: {id: 1},
    };
    const {toJSON} = render(<FullScorecard route={route} />);
    expect(toJSON()).toMatchSnapshot();
  });
  it('should render screen title', () => {
    const route = {
      params: {id: 1},
    };
    render(<FullScorecard route={route} />);
    expect(screen.getByText('Scorecard')).toBeTruthy();
  });
  it('should display back button', () => {
    const route = {
      params: {id: 1},
    };
    render(<FullScorecard route={route} />);
    expect(screen.getByText('Back')).toBeTruthy();
  });
  it('should navigate backk to scorecards screen', () => {
    const route = {
      params: {id: 1},
    };
    const navigation = {
      navigate: jest.fn(),
    };
    const {getByText} = render(
      <FullScorecard route={route} navigation={navigation} />,
    );
    fireEvent.press(getByText('Back'));
    expect(navigation.navigate).toHaveBeenCalledWith('Scorecard');
  });
});
