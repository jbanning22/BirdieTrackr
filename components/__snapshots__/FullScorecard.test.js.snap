// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full Scorecard Screen should match the snapshot 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#DDDDDD",
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <Text
    style={
      {
        "color": "#DB6F52",
        "fontFamily": "Helvetica",
        "fontSize": 44,
        "fontWeight": "400",
        "textAlign": "center",
      }
    }
  >
     Scorecard
  </Text>
  <View
    style={
      {
        "flex": 0,
        "width": "100%",
      }
    }
  >
    <RCTScrollView
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      pagingEnabled={true}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsHorizontalScrollIndicator={false}
      snapToAlignment="center"
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View />
    </RCTScrollView>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "alignSelf": "center",
          "marginTop": 50,
        }
      }
    >
      Back
    </Text>
  </View>
</RNCSafeAreaView>
`;
